GX-TimeFrequency/test/conftest.py

```diff
@@ -5,50 +5,47 @@
 
 
 def pytest_addoption(parser):
+    """ Custom command-line options parsed added to pytest.
     """
-    Custom command-line options parsed added to pytest.
-    Function used by pytest.
-
-    These are ONLY used in the test runner to determine the full path to
-    the fortran executable and the run settings.
-    """
-
     parser.addoption('--root',
                      type=str,
                      dest='root',
                      required=False,
                      help='Full path for GreenX root'
                      )
 
+    parser.addoption('--build-dir',
+                     type=str,
+                     dest='build_dir',
+                     default=None,
+                     required=False,
+                     help='Full path to build directory'
+                     )
+
     parser.addoption('--binary',
                      type=str,
                      dest='binary',
                      required=False,
-                     help='Name of the executable with full path prepending'
+                     help='Name of the executable with full path prepended'
                      )
 
 
-# Define metafunc entries from command-line arguments.
-# metafunc can be passed to pytest fixtures.
-def pytest_generate_tests(metafunc):
-    """ Add parameter or parameters to metafunc.
-
-    :param metafunc
-    """
-    for option in ['root', 'binary']:
-        if option in metafunc.fixturenames:
-            metafunc.parametrize(option, metafunc.config.getoption(option))
-
-
 @pytest.fixture(scope="session")
-def greenx_build_root() -> str:
-    """ Get GreenX build directory from an environment variable
-    set by CMake, during the build process.
+def greenx_build_root(request) -> str:
+    """ Get GreenX build directory.
 
     :return: Environment variable string
     """
-    ENV_VAR = 'GX_BUILD_DIR'
-    return os.getenv(ENV_VAR)
+    # First, try from command line arg.
+    build_dir = request.config.getoption('build_dir')
+    if build_dir:
+        if not Path(build_dir).is_dir():
+            raise NotADirectoryError(f'{build_dir} is not a directory')
+        return build_dir
+
+    #  If this is not passed, look for an environment variable
+    if build_dir is None:
+        return os.getenv('GX_BUILD_DIR')
```

Comment:
This can return None, which is inconsistent with the function's type hint.

I recommend either to raise an error or `pytest.skip()` when tests want the build dir and it isn't known.

(I would also always return is at a Path object rather than string)

Reply:
If the directory is None, the tests can't run so I don't think `pytest.skip() `is appropriate.
Changes implemented. Let me know if they're ok. More branching logic than I would have liked, but the number of options for obtaining the path won't grow, so I think it's fine